### PR TITLE
init

### PR DESCRIPTION
**Pull request by Wescoeur**
@benjamreis For the moment you can ignore the `qcow2-stream.c` file. I must factorize it and fix a bug with a 2097152-bytes clusters image.

### CODE REVIEW

File: tools/stream-to-file.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "xcp-ng/vdi-stream.h"
+
+// =============================================================================
+
+int main (int argc, char *argv[]) {
+  if (argc < 4) {
+    fprintf(stderr, "Usage: %s <output> <format> <vdi> [base]\n", *argv);
```
Comment (benjamreis) [inline tools/stream-to-file.c:29]: Why base is in between "[]" and the other "<>"?
Does it mean there's optionnal argument? In this case the parsing should be smarter.

Why not --help to print usage?

File: tools/dump-info.c [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "xcp-ng/vdi-stream.h"
+
+// =============================================================================
+
+int main (int argc, char *argv[]) {
+  if (argc < 3) {
+    fprintf(stderr, "Usage: %s <format> <vdi>\n", *argv);
```
Comment (benjamreis) [inline tools/dump-info.c:28]: Why not --help to print usage?

File: tools/stream-to-file.c [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    goto fail;
+  }
+
+  if (xcp_vdi_stream_open(stream, argv[2], argv[3], argc >= 5 ? argv[4] : NULL) < 0) {
+    fprintf(stderr, "Unable to open stream because: `%s`.\n", xcp_vdi_stream_get_error_string(stream));
+    goto fail;
+  }
+
+  if (!(output = fopen(argv[1], "wb"))) {
+    fprintf(stderr, "Unable to open `%s`.\n", argv[1]);
```
Comment (benjamreis) [inline tools/stream-to-file.c]: Print errno?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+cd build
+cmake .. -DCMAKE_PREFIX_PATH=<path_to_xcp-ng-generic-build_dir>
+make
+```
+
+## Tools
+
+Two tools linked to this library are provided:
+- `dump-info` to extract metadata of an image
+- `stream-to-file` to stream an image chain to a file
```
Comment (benjamreis) [inline README.md:26]: Add usage and perhaps example?

File: src/image-format/qcow2.c [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+    TAILQ_REMOVE(&cache->sortedEntries, entry, sortedEntry);
+  }
+
+  entry->l2TableOffset = l2TableOffset;
+  TAILQ_INSERT_TAIL(&cache->entries[hash], entry, entry);
+  TAILQ_INSERT_HEAD(&cache->sortedEntries, entry, sortedEntry);
+
+  const XcpError ret = xcp_fd_pread(image->fd, entry->l2Table, clusterSize, (off_t)l2TableOffset);
+  if (ret == XCP_ERR_ERRNO) {
+    set_error(error, "Unable to read L2 table at offset %# in %s" PRIx64, l2TableOffset, image->filename);
```
Comment (benjamreis) [inline src/image-format/qcow2.c]: because: errno?

File: src/image-format/qcow2.c [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+
+  // 3. Read backing filename.
+  if (header->backingFileOffset) {
+    const uint32_t size = header->backingFileSize;
+    if (size >= XCP_MIN(image->clusterSize - header->backingFileOffset, sizeof image->backingFile)) {
+      set_error(error, "Backing filename is so big");
+      goto fail;
+    }
+    const int ret = (int)xcp_fd_pread(image->fd, image->backingFile, size, (off_t)header->backingFileOffset);
+    if (ret < 0) {
```
Comment (benjamreis) [inline src/image-format/qcow2.c]: Why not compare to XCP_ERRNO ?

File: src/vdi-stream.c [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+void xcp_vdi_stream_dump_info (const XcpVdiStream *stream, int fd) {
+  if (stream->driver)
+    (*stream->driver->dumpInfo)(stream, fd);
+}
+
+static void xcp_vdi_stream_co_read_wrapper (void *userData) {
+  XcpVdiStream *stream = (XcpVdiStream *)userData;
+  (stream)->streamBuf->coRet = (*stream->driver->read)(stream);
```
Comment (benjamreis) [inline src/vdi-stream.c]: why stream is between "()"?

**Review by benjamreis — CHANGES_REQUESTED** [review]

File: tools/stream-to-file.c (hunk 1 above)
Comment (Wescoeur) [reply tools/stream-to-file.c:29]: https://en.wikipedia.org/wiki/Usage_message#Pattern

I don't understand what you mean by "parsing should be smarter".
If you know the unix message pattern, I don't see a good idea to add a `--help` param, this tool is very small.

File: tools/dump-info.c (hunk 2 above)
Comment (Wescoeur) [reply tools/dump-info.c:28]: Same idea here, it's a small tool.

File: src/vdi-stream.c (hunk 7 above)
Comment (Wescoeur) [reply src/vdi-stream.c]: No reason. ;)

**Comment by Wescoeur** [comment]
@benjamreis `README.md` has been updated, `set_error` is now more secure (format is checked), and `errno` is used in some places.

**Review by benjamreis — APPROVED** [review]

**Pull request merged**